### DISCUSSION TITLE
xhciep: Transfer timed out with Circuitpython device

### DISCUSSION THREAD

**Discussion by diyelectromusic** [rsta2/circle#460] [Q&A] [answered]
I'm using a Circuitpython v8 USB MIDI device plugged into a Pi 4 running Circle (MiniDexed) which seems to be recognised as the following USB devices:
int2-2-0
inta-0-0
int8-6-50
int1-1-0
int1-3-0

The only one I'm really interested in is int1-3-0 for audio/MIDI, but when accessing the SD card filesystem (this is on MiniDexed) for some reason I'm getting xhciep Transfer timeouts which causes the execution/UI to pause (presumably whilst the timeout occurs).

I've tried using usbignore= for some of the above - with int8-6-50 being the most suspect as a mass storage device, but it doesn't seem to make a difference.

I've also used the following boot.py code ([more here](https://learn.adafruit.com/customizing-usb-devices-in-circuitpython/circuitpy-midi-serial)):
```
import usb_hid, usb_midi
usb_hid.disable()
usb_midi.enable()
```

And also tried disabling storage completely too using:
```
import storage
storage.disable_usb_drive()
```
(which I know it warns not to do - I still had REPL access so could turn it back on again).

But so far, it doesn't seem to make any difference.  As I'm largely trying things pretty much at random at the moment, I just wondered if anyone had any more concrete ideas for what might be going on?

I'll keep trying various configurations and see if anything changes.

Thanks,
Kevin

**Comment by rsta2** [comment]
I'm trying to reproduce this with *sample/29-miniorgan* and a RPi Pico. What would be helpful is a simple *code.py* for Circuitpython, which sends some MIDI events (notes) to the USB host. Unfortunately I'm not familiar with Circuitpython.

**Reply by diyelectromusic** [reply]
Ah great.  In that case I'll try to do the same, but it won't be until later today.

I think this is probably the simplest Circuitpython that will play a MIDI note over MIDI USB.
```
from time import sleep
import board
import digitalio
import usb_midi
import adafruit_midi
from adafruit_midi.note_off import NoteOff
from adafruit_midi.note_on import NoteOn

led = digitalio.DigitalInOut(board.LED)
led.direction = digitalio.Direction.OUTPUT

usb_midi = adafruit_midi.MIDI(
    midi_out=usb_midi.ports[1],
    out_channel=0   # Channel 1
    )

while True:
    usb_midi.send(NoteOn(64,127))
    led.value = True
    sleep(0.5)
    usb_midi.send(NoteOff(64,0))
    led.value = False
    sleep(0.5)
```

This needs the adafruit_midi library folder in the /lib folder (which can be grabbed from the Adafruit Library collection).  I don't think it needs anything else.

Kevin

**Reply by rsta2** [reply]
*sample/29-miniorgan* works successfully here with Circuitpython 9.0.5 installed on a RPi Pico and this file *code.py*:

```
import usb_midi
import time

midi_out = usb_midi.ports[1]

while 1:
	note = 40
	while note < 70:
		note_on = bytes([0x90, note, 127])
		note_off = bytes([0x80, note, 0])
		note = note + 1
		midi_out.write(note_on)
		time.sleep(0.2)
		midi_out.write(note_off)
		time.sleep(0.5)
```

I have no file *boot.py* on the Circuitpython drive.

Rene

**Reply by diyelectromusic** [reply]
It was causing issues when changing performances in MiniDexed, so I took that as probably related to the SD card?  MIDI was working fine, which is why I thought it perhaps a side effect of one of the other USB endpoints.

As I say, I'll see if I can get something down to a smaller test case, but not really knowing what it is relating to, it won't be easy... I really need a trace through the USB driver to know what transfer was triggering the timeout.

I was using a Circuitpython 8 device, so that is certainly something to try first - I'll grab the latest :)

Thanks,
Kevin

**Reply by rsta2** [reply]
With this patch it tells you the EP number, which causes the timeout message:

```
diff --git a/lib/usb/xhciendpoint.cpp b/lib/usb/xhciendpoint.cpp
index 449f1d2..ab6097a 100644
--- a/lib/usb/xhciendpoint.cpp
+++ b/lib/usb/xhciendpoint.cpp
@@ -223,7 +223,8 @@ boolean CXHCIEndpoint::Transfer (CUSBRequest *pURB, unsigned nTimeoutMs)
        {
                if (CTimer::Get ()->GetTicks () - nStartTicks >= HZ)
                {
-                       CLogger::Get ()->Write (From, LogDebug, "Transfer timed out");
+                       CLogger::Get ()->Write (From, LogDebug, "Transfer timed out (ep %u)",
+                                               (unsigned) pURB->GetEndpoint ()->GetNumber ());
 #ifdef XHCI_DEBUG
                        m_pDevice->DumpStatus ();
 #endif
```

You can use `lsusb -v` on Linux to search for this endpoint number for the Circuitpython device. The direction bit (0x80) is not included in the shown EP number.

**Reply by diyelectromusic** [reply]
Ah brilliant!  Many thanks - will add that in and see what it does! :)

**Reply by diyelectromusic** [reply, marked as answer]
Ok.  So, turns out it is the MIDI int1-3-0 endpoint - so the one I actually want.  Then I realised this is happening on send - i.e. when MiniDexed tries to send some data back over the MIDI link.

Also, it turns out it is a lot worse when changing voices.  Then I realised what is going on...

In MiniDexed there is an option someone added (that personally I don't agree with!) that automatically does a SysEx voice dump of all voice parameters whenever the voice changes.  This happens when a performance is selected too, as it resets voices on each TG, but performances have to be selected to load, but voices change on menu changing - so every click of the Voice menu is dumping data out on all USB MIDI interfacces!

But the Circuitpython device isn't listening for USB data, it only sends.

So I think the timeout is because MiniDexed is spamming the USB MIDI interface and the CP device is ignoring it...

Testing again with the option to disable the auto-sending of the SysEx data and yep - all fine!?

Sorry for wasting your time, but also thanks for that clue - without that link I suspect it would have taken a lot longer to work out what was going on!

Time to update the MiniDexed wiki I think :)

Kevin

**Reply by rsta2** [reply]
I see. No problem, I will probably generally add this patch.

Rene
